dashboard/app/controllers/projects_controller.rb

```diff
@@ -310,19 +310,37 @@ def create_new
     )
   end
 
-  # GET /projects(/script/:script_id)/level/:level_id
+  # GET /projects(/script/:script_id)/level/:level_id(/user/:user_id)
   # Given a level_id and the current user (or signed out user), get the existing project
   # or create a new project for that level and user. If a script_id is provided, get or
   # create the project for that level, script and user
   # Returns json: {channel: <encrypted-channel-token>}
   def get_or_create_for_level
     script_id = params[:script_id]
+    user_id = params[:user_id]
     level = Level.find(params[:level_id])
+
+    # If viewing another user's work, ensure that we have permission.
+    if user_id
+      script_level = level.script_levels.find_by_script_id(script_id)
+      user = User.find(user_id)
+      unless can?(:view_as_user, script_level, user)
+        return render(status: :forbidden, json: {error: "Access denied."})
+      end
+    end
+
     error_message = under_13_without_tos_teacher?(level)
     return render(status: :forbidden, json: {error: error_message}) if error_message
+
     # get_storage_id works for signed out users as well, it uses the cookie to determine
     # the storage id.
-    user_storage_id = get_storage_id
+    user_storage_id = if user_id
+                        # todo: handle this not finding a valid storage id
```

Comment:
Should we fall back to the current user's channel ID in this case? Or throw an error? (Curious what we do for non-lab2 labs in this situation)

Reply:
Pulling on this thread led to some new work that has arrived in commit https://github.com/code-dot-org/code-dot-org/pull/58269/commits/76b5dbc623485032bc51cf44b6e85b960f0b79e8.  With that work, we:
- no longer create a channel if one doesn't exist when a teacher is looking at a student's work;
- update the blockly workspace to be readonly or not depending on whether a teacher is viewing a student's work or their own;
- handle the situation in which a student has not yet done any work yet, in which case the starter code is shown.